vortex-array/src/stats/bound.rs

```diff
@@ -0,0 +1,312 @@
+        self.0
+    }
+}
+
+impl<T: PartialOrd + Clone> StatBound<T> for UpperBound<T> {
+    fn lift(value: Precision<T>) -> Self {
+        Self(value)
+    }
+
+    /// The meet or tightest covering bound
+    fn union(&self, other: &Self) -> Option<UpperBound<T>> {
+        Some(UpperBound(match (&self.0, &other.0) {
+            (Exact(lhs), Exact(rhs)) => Exact(max(lhs, rhs)?.clone()),
+            (Inexact(lhs), Inexact(rhs)) => Inexact(max(lhs, rhs)?.clone()),
+            (Inexact(lhs), Exact(rhs)) => {
+                if rhs >= lhs {
+                    Exact(rhs.clone())
+                } else {
+                    Inexact(lhs.clone())
+                }
+            }
+            (Exact(lhs), Inexact(rhs)) => {
+                if rhs <= lhs {
+                    Exact(lhs.clone())
+                } else {
+                    Inexact(rhs.clone())
+                }
+            }
+        }))
+    }
+
+    fn intersection(&self, other: &Self) -> Option<JoinResult<UpperBound<T>>> {
+        Some(match (&self.0, &other.0) {
+            (Exact(lhs), Exact(rhs)) => {
+                if lhs == rhs {
+                    JoinResult::Join(UpperBound(Exact(lhs.clone())))
+                } else {
+                    // The two intervals do not overlap
+                    JoinResult::None
+                }
+            }
+            (Inexact(lhs), Inexact(rhs)) => {
+                JoinResult::Join(UpperBound(Inexact(min(lhs, rhs)?.clone())))
+            }
+            (Inexact(lhs), Exact(rhs)) => {
+                if rhs <= lhs {
+                    JoinResult::Join(UpperBound(Exact(rhs.clone())))
+                } else {
+                    // The two intervals do not overlap
+                    JoinResult::None
+                }
+            }
+            (Exact(lhs), Inexact(rhs)) => {
+                if rhs >= lhs {
+                    JoinResult::Join(UpperBound(Exact(lhs.clone())))
+                } else {
+                    // The two intervals do not overlap
+                    JoinResult::None
+                }
+            }
+        })
+    }
+}
+
+impl<T: PartialOrd> PartialEq<T> for UpperBound<T> {
+    fn eq(&self, other: &T) -> bool {
+        match &self.0 {
+            Exact(lhs) => lhs == other,
+            _ => false,
+        }
+    }
+}
+
+// We can only compare exact values with values and Precision::inexact values can only be greater than a value
+impl<T: PartialOrd> PartialOrd<T> for UpperBound<T> {
+    fn partial_cmp(&self, other: &T) -> Option<Ordering> {
+        match &self.0 {
+            Exact(lhs) => lhs.partial_cmp(other),
+            Inexact(lhs) => lhs.partial_cmp(other).and_then(|o| {
+                if o == Ordering::Greater {
+                    None
+                } else {
+                    Some(o)
+                }
+            }),
+        }
+    }
+}
+
+#[inline]
+pub fn min<T: PartialOrd>(a: T, b: T) -> Option<T> {
```

Comment:
I think you should just implement this properly.

```
pub trait PartialMin<T = Self>: PartialOrd<T> {
  fn partial_min(&self, other: &T) -> Option<T>;
}

impl<T: PartialOrd> PartialMin for T {}

```

Reply:
This is what the std lib does. I can add I just never see the point of these traits.

Reply:
Makes the functions accessible on the struct (mroe discoverable), and allows you to add the trait constraint to generics